### PR TITLE
Using CUDA.jl

### LINKED ISSUE #96: Drop CuArrays.jl from deps and move to CUDA.jl
CuArrays has Adapt ^1 as its dep while Plots requires Adapt ^2
This creates some trouble. E.g., if you have Plots v1.6 installed and you try
```julia
]add GeophysicalFlows
```
you end up with `GeophysicalFlows v0.3.0`.

### PR DESCRIPTION
**Pull request by navidcy**
- Drops CuArrays.jl and CUDAapi.jl and uses CUDA.jl instead
- Disallows scalar operations on GPU

The fact that I had to use `@CUDA.allowscalar` so many times (especially in the `MultiLaywerQG`) probably means something. It seems like `@views` use `getindex` on the GPU; I was getting errors if I didn't include `@CUDA.allowscalar` in front of expressions with `@views`.

**!!** This should **only** be merged after https://github.com/FourierFlows/FourierFlows.jl/pull/198 is in master. **!!** 

Closes #96

### CODE REVIEW

File: src/barotropicqgql.jl [hunk 1]
```diff
@@ -201,11 +202,11 @@ end
 # -------
 
 function calcN_advection!(N, sol, t, clock, vars, params, grid)
-  Kr = [ grid.kr[i] for i=1:grid.nkr, j=1:grid.nl]
+  Kr = CUDA.@allowscalar [ grid.kr[i] for i=1:grid.nkr, j=1:grid.nl ]
   @. vars.zetah = sol
-  @. vars.zetah[Kr .== 0] = 0
+  CUDA.@allowscalar @. vars.zetah[Kr .== 0] = 0
```
[the remark this answers is not in the document]
Comment (navidcy) [reply src/barotropicqgql.jl:207]: Damn... This is back from my much-more-naïve days.

Sure -- I agree.

Options:

- Define it within the module and let `calcN!` to get it from module's global scope? --> slower?
- Give it as part of `Params`? --> but it's not a paramater... it's a derived `grid` quantity...

File: src/barotropicqgql.jl [hunk 2]
```diff
@@ -201,11 +202,11 @@ end
 # -------
 
 function calcN_advection!(N, sol, t, clock, vars, params, grid)
-  Kr = [ grid.kr[i] for i=1:grid.nkr, j=1:grid.nl]
+  Kr = CUDA.@allowscalar [ grid.kr[i] for i=1:grid.nkr, j=1:grid.nl ]
```
[the remark this answers is not in the document]
Comment (navidcy) [reply src/barotropicqgql.jl:205]: Damn... This is back from my much-more-naïve days.

Sure -- I agree.

Options:

- Define it within the module and let `calcN!` to get it from module's global scope? --> slower?
- Give it as part of `Params`? --> but it's not a paramater... it's a derived `grid` quantity...

[the remark this answers is not in the document]
Comment (navidcy) [reply src/barotropicqgql.jl]: Now with `@. zeta[:, 1] = 0`, ... we don't even need it.

**Comment by navidcy** [comment]
> Using `@views` is certainly just a convenience. We'll get better performance writing kernels. This is also true even in cases where we do not use `@views` (kernels seem to be faster than naive broadcasting, at least right now, which may be an inefficiency in broadcasting that could be fixed sometime in the future). Note that if we write kernels and use `KernelAbstractions` we also will get multithreaded speed up even on algebraic operations (not just FFTs).

OK, I see your point. It takes a bit away from the "Julia can seamlessly be GPU-ready" idea that I had in mind. But, what you are saying is that if the developers work a bit harder here (writing kernels) then the experience will be seamless for the users, right?
But then what's then this is the same for `PyCUDA` and things like that, no?

**Comment by navidcy** [comment]
@glwagner don't merge this. First FourierFlows/FourierFlows.jl#198 needs to be merged, make a new release of FourierFlows.jl and then we should remove `Manifest.toml` from here and add a `[compat]` entry in `Project.toml` to force GeophysicalFlows.jl use the latest release of FourierFlows.jl.

**Comment by glwagner** [comment]
> OK, I see your point. It takes a bit away from the "Julia can seamlessly be GPU-ready" idea that I had in mind. But, what you are saying is that if the developers work a bit harder here (writing kernels) then the experience will be seamless for the users, right?
But then what's then this is the same for PyCUDA and things like that, no?

I think I was confusing. If we write kernels, they will be multithreaded and possibly will have _slightly_ better GPU performance than broadcasting.

Note that fuse broadcasts could become multithreaded in the future. They just aren't right now. I think GPU broadcasting could be improved, perhaps.

I am not implying that hand-written kernels are _necessary_ for performant code. I apologize if I implied that.

**Pull request merged**